### DISCUSSION TITLE
Can I get the original object of `.extend()`?

### DISCUSSION THREAD

**Discussion by j-murata** [colinhacks/zod#2206] [Q&A] [answered]
For example, [in this case](https://github.com/colinhacks/zod#extend), does `DogWithBreed` contain the information "extended from `Dog`"?

I would like to output the following TypeScript typedef from these schema objects:

```js
import { z } from "zod";

const Dog = z.object({
    name: z.string(),
    age: z.number(),
});

const DogWithBreed = Dog.extend({
    breed: z.string(),
});
```
↓ ↓ ↓
```ts
interface Dog {
    name: string;
    age: number;
}

interface DogWithBreed extends Dog {
    breed: string;
}
```

**Comment by j-murata** [comment]
If I use `.and()` instead of `.extend()`, I see that the information remains in `left` and `right` under `_def`.

```js
const DogWithBreed = Dog.and(z.object({
    breed: z.string(),
}));
```

In the case of `.extend()`, does the information on the inheritance source disappear completely...

**Reply by kostysh** [reply]
Is this what you are looking for?

```typescript
type DogType = z.infer<typeof Dog>;

// type DogType = {
//     name: string;
//     age: number;
// }
```

```typescript
type DogWithBreedType = z.infer<typeof DogWithBreed>;

// type DogWithBreedType = {
//     name: string;
//     age: number;
//     breed: string;
// }
```

**Reply by j-murata** [reply]
@kostysh Thank you for your reply.
Sorry, that is not what I am looking for.

I want to keep the TypeScript type information separate from the schema object, and most importantly, even with your suggested method, the information that `DogWithBreedType` inherits from `DogType` is lost, right?

**Reply by kostysh** [reply]
It is not possible to infer the type from the schema in the form of:

```typescript
interface DogWithBreed extends Dog {
    breed: string;
}
```

you always will get the resulting type that will be a combination of parent and child types.
And finally, technically there is no difference between the `extends Dog..` version and

```typescript
type DogWithBreedType = z.infer<typeof DogWithBreed>;

// type DogWithBreedType = {
//     name: string;
//     age: number;
//     breed: string;
// }
```

**Reply by j-murata** [reply]
I guess that's true.

It is not OK if the final result has all the keys. It is important to be able to show the schema user "what type it inherits from".

I will try to see if I can present an inheritance relationship between schemas using `.and()`. Thanks.

**Reply by kostysh** [reply]
Of course, you can use `.and()` to be able to see inferred types like `{ prop: string; } & { prop2: number; }`. From a typescript type point of view, there is no difference from `{ prop: string; prop2: number }`.

But, using `.and()` you should remember that, for example, `{ prop: string; } & { prop: number; }` produces a type `{ prop: never; }` that is sometimes not so obvious if your schemas are quite large.

I recommend you do not use `.and()` for inheritance.

**Reply by j-murata** [reply]
As I have said many times, even if you define the types as follows, Schema2 users will not know whether `prop` is a member inherited from Schema1 or not (unless they bother to compare it with Schema1), which is not good for usability. The more complex the schema, the greater the effort involved.
```ts
type Schema1 = { prop: string };
type Schema2 = { prop: string, prop2: number };
```

If the schema was defined with `.and()`, I believe I can show them the common and differential members from the `left` and `right` information in its `_def`.
I wonder why `.extend()` is a simple overwrite implementation.

**Comment by j-murata** [comment, marked as answer]
As you can see in the thread discussion above, I tried to use `.and()` to output an inherited TS interface, but decided against this method because the return value of `.and()` is not a `ZodObject` and the deeper the nesting, the more difficult it is to handle.
However, the original `.extend()` does not preserve the inheritance source information, so I decided to override `.extend()` as follows. 
Using this and [`zod-to-ts`](https://www.npmjs.com/package/zod-to-ts), I can now properly output the inheritance relationship such as `interface DogWithBreed extends Dog`.
```js
import { ZodObject } from 'zod';
{
    const originalExtend = ZodObject.prototype.extend;
    ZodObject.prototype.extend = function (...args) {
        const returnValue = originalExtend.call(this, ...args);
        returnValue._def._extends = {
            parent: this,
            props: args[0],
        };
        return returnValue;
    }
}
```

**Comment by j-murata** [comment]
Now that my problem has been resolved, I will close this discussion.
